locallib.php

```diff
@@ -628,6 +631,272 @@ private function process_action_manual_allocation() {
+            if (empty($choicessortedwithgroupscount[$i])) {
+                continue;
+            }
+            $choicessorted = array_merge($choicessorted, $choicessortedwithgroupscount[$i]);
+        }
+        if (!empty($choicessortedwithgroupscount[0])) {
+            $choicessorted = array_merge($choicessorted, $choicessortedwithgroupscount[0]);
+        }
+
+        // This is kind of a dilemma. We want the choices to be filled up beginning at the one with the least places left to fill it
+        // up as quickly as possible.
+        // However, in case of group restrictions this will lead to problems as we are assigning users which have to be assigned to
+        // specific choices (because all others cannot be assigned to it). So in the end these choices will not be available when
+        // we arrive at the choice with the group restrictions.
+        // Therefore, we are first filling up choices with group restrictions first (beginning at choices with fewer groups). Only
+        // in case we have the same amount of groups for two choices or we have no group restrictions at all we pick choices with
+        // fewer places left first (see foreach loop with usort a few lines above).
+
+        return !empty($choicessorted) ? array_shift($choicessorted) : -1;
+    }
+
+    /**
+     * Try to distribute all currently unallocated users.
+     *
+     * @param string $distributionalgorithm the distributionalgorithm which should be used, you can choose between
+     *  ACTION_DISTRIBUTE_UNALLOCATED_EQUALLY and ACTION_DISTRIBUTE_UNALLOCATED_FILL
+     * @return void
+     * @throws dml_exception
+     */
+    public function distribute_users_without_choice(string $distributionalgorithm): void {
+        // This could need some extra memory, especially because we are caching some data structures in memory while
+        // running the algorithm.
+        raise_memory_limit(MEMORY_EXTRA);
+        core_php_time_limit::raise();
+
+        // This will retrieve a list of users we are trying to assign to choices, sorted from the least group membership count to
+        // more group memberships. Users without group memberships will be at the end of the array.
+        $possibleusers = $this->get_undistributed_users();
+
+        $transaction = $this->db->start_delegated_transaction();
+
+        $usertoassign = array_shift($possibleusers);
+        // As long as we have a user to assign, we try to assign him.
+        while ($usertoassign != null) {
+
+            // Calculate the choice to assign the user to depending on the given algorithm.
+            $choicetoassign = $this->get_next_choice_to_assign_user($distributionalgorithm, $usertoassign);
+            if ($choicetoassign === -2) {
+                // This means there are no free places left in any choice for any user, so we can stop the algorithm
+                // as a whole.
+                return;
```

Comment:
There should be a break here or alternatively add `$transaction->allow_commit()` before the return statement otherwise the made allocations are not registered